app/controllers/api/account.php

```diff
@@ -1647,6 +1654,99 @@
         throw new Exception('Session not found', 404);
     });
 
+App::patch('/v1/account/sessions/:sessionId/oauth2-tokens')
+    ->desc('Update OAUth2 Tokens')
+    ->groups(['api', 'account'])
+    ->label('scope', 'account')
+    ->label('event', 'account.sessions.update')
+    ->label('sdk.auth', [APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_JWT])
+    ->label('sdk.namespace', 'account')
+    ->label('sdk.method', 'updateOAuth2Tokens')
+    ->label('sdk.description', '/docs/references/account/update-oauth2-tokens.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_SESSION)
+    ->label('abuse-limit', 10)
+    ->param('sessionId', null, new UID(), 'Session ID. Use the string \'current\' to update the current device session.')
+    ->param('force', false, new Boolean(), 'Should generate new token even if current one is still valid?', true)
+    ->inject('request')
+    ->inject('response')
+    ->inject('user')
+    ->inject('dbForProject')
+    ->inject('project')
+    ->inject('locale')
+    ->inject('audits')
+    ->inject('events')
+    ->inject('usage')
+    ->action(function ($sessionId, $force, $request, $response, $user, $dbForProject, $project, $locale, $audits, $events, $usage) {
+        /** @var Appwrite\Utopia\Request $request */
+        /** @var boolean $force */
+        /** @var Appwrite\Utopia\Response $response */
+        /** @var Utopia\Database\Document $user */
+        /** @var Utopia\Database\Database $dbForProject */
+        /** @var Utopia\Database\Document $project */
+        /** @var Utopia\Locale\Locale $locale */
+        /** @var Appwrite\Event\Event $audits */
+        /** @var Appwrite\Event\Event $events */
+        /** @var Appwrite\Stats\Stats $usage */
+
+        $sessionId = ($sessionId === 'current')
+            ? Auth::sessionVerify($user->getAttribute('sessions'), Auth::$secret)
+            : $sessionId;
+
+        $sessions = $user->getAttribute('sessions', []);
+
+        foreach ($sessions as $key => $session) {/** @var Document $session */
```

Comment:
I can definitely see your problem with this. We could try and optimize the code, but could also consider deprecating the sessions attribute and just use a subquery like we do for the project sub-entities.

Reply:
I checked the rest of the code and I believe we currently do a lot of duplicate work. Every write is done on both 'sessions' and 'user'. We 100% should have subQuery on sessions, but that would need to be a separate refactor.

With that said, I would stick with this solution that can get slower with many active sessions, and fix it alongside other endpoints once we introduce subQuery for sessions.

Reply:
Agree, and it will be better to address it in a new PR.